src/daft-image/src/ops.rs

```diff
@@ -359,3 +546,339 @@ pub fn fixed_image_html_value(arr: &FixedShapeImageArray, idx: usize, truncate:
+            let left_pixel = pixel_data[y * 9 + x];
+            let right_pixel = pixel_data[y * 9 + x + 1];
+            hash_bits.push(if left_pixel < right_pixel { '1' } else { '0' });
+        }
+    }
+
+    let hash = hash_bits.into_iter().collect::<String>();
+
+    // Validate length (safety check)
+    if hash.len() != 64 {
+        return Err(DaftError::ValueError(format!(
+            "Difference hash should be 64 characters, got {}",
+            hash.len()
+        )));
+    }
+
+    Ok(hash)
+}
+
+#[allow(clippy::needless_range_loop)]
+fn compute_wavelet_hash(img: CowImage) -> DaftResult<String> {
+    // Convert to grayscale
+    let gray_img = img.into_mode(daft_schema::prelude::ImageMode::L);
+
+    // Resize to 64x64 for wavelet transform
+    let resized = gray_img.resize(64, 64);
+
+    // Get pixel data
+    let pixel_data = resized.as_u8_slice();
+
+    // Convert to float matrix for wavelet transform
+    let mut matrix = vec![vec![0.0f64; 64]; 64];
+    for y in 0..64 {
+        for x in 0..64 {
+            matrix[y][x] = pixel_data[y * 64 + x] as f64;
+        }
+    }
+
+    // Apply Haar wavelet transform
+    let mut coeffs = matrix;
+    let size = 64;
+
+    // Apply wavelet transform iteratively
+    let mut current_size = size;
+    while current_size > 1 {
+        // Create a temporary copy for the transform
+        let temp = coeffs.clone();
+
+        // Horizontal transform
+        for y in 0..current_size {
+            for x in 0..current_size / 2 {
+                let left = temp[y][x * 2];
+                let right = temp[y][x * 2 + 1];
+
+                // Average (low frequency)
+                coeffs[y][x] = f64::midpoint(left, right);
+                // Difference (high frequency)
+                coeffs[y][x + current_size / 2] = (left - right) / 2.0;
+            }
+        }
+
+        // Vertical transform
+        for x in 0..current_size {
+            for y in 0..current_size / 2 {
+                let top = coeffs[y * 2][x];
+                let bottom = coeffs[y * 2 + 1][x];
+
+                // Average (low frequency)
+                coeffs[y][x] = f64::midpoint(top, bottom);
+                // Difference (high frequency)
+                coeffs[y + current_size / 2][x] = (top - bottom) / 2.0;
+            }
+        }
+
+        current_size /= 2;
+    }
+
+    // Extract 8x8 top-left corner (low frequency components)
+    let mut hash_bits = Vec::new();
+
+    // Calculate median of the 8x8 block
+    let mut values = Vec::new();
+    for y in 0..8 {
+        for x in 0..8 {
+            values.push(coeffs[y][x]);
+        }
+    }
+    values.sort_by(|a, b| a.partial_cmp(b).unwrap());
+    let median = if !values.is_empty() {
+        values[values.len() / 2]
+    } else {
+        0.0
+    };
+
+    // Generate hash bits
+    for y in 0..8 {
+        for x in 0..8 {
+            hash_bits.push(if coeffs[y][x] > median { '1' } else { '0' });
+        }
+    }
```

Comment:
It appears that all of these hash functions are essentially a `u64` output. We could instead use bitwise operations and have the return type be `DataType::Uint64` instead of `String`.


```rs
    let mut hash: u64 = 0;
    for y in 0..8 {
        for x in 0..8 {
            let bit_position = y * 8 + x;
            if coeffs[y][x] > median {
                hash |= 1u64 << bit_position;
            }
        }
    }
    Ok(hash)
```